### PR TITLE
Fix concurrency issues

### LINKED ISSUE #157: Race conditions
It seems like there are a few serverside race condition issues that can pop up if multiple requests are made at the same time.

For example, if I upload a box, a few things happen:

1. A new `Box` record gets created on the database
1. The new box's ID gets added to the owner's `user._orderedBoxIds` array. This is used to ensure that the boxes have a consistent order, which can be arbitrarily changed by the user.

More concretely, with waterline step 2 looks something like this:

```js
function addIdToArray (ownerName, id) {
  return User.findOne({name: ownerName}).then(user => {
    user._orderedBoxIds.push(id);
    return user.save();
  });
}
```

The problem is that this step requires two distinct database queries under the hood -- one to get the current list of IDs, and one to update the list with the new ID. This makes it fairly easy to cause (and reproduce) a race condition bug if two boxes are added simultaneously, because the unmodified array is received by the first query for both requests, and then a different value is added in the second query for each request. This causes one of the IDs to get lost.

It seems like MongoDB has a `$push` operator, which could be used to directly add the ID in a single query, i.e.

```js
db.user.update({name: ownerName}, {$push: {_orderedBoxIds: id}})
```

...but it's not supported by waterline. If it would solve the issue, we could probably just run it natively and bypass waterline in this case.

### PR DESCRIPTION
**Pull request by not-an-aardvark**
Fixes #157

tl;dr We should avoid using waterline's `.save()` method for database models, and instead use .update() to make updates to the database. If a document is modified between the time that it is fetched and `.save()` is called, the changes will be overwritten by `.save()` but not by `.update()`.

There is still the possibility of leaving the database in an inconsistent state if connection is randomly lost (which is sort of difficult to avoid since mongo doesn't have transactions), but there shouldn't be any more race condition issues now.

### CODE REVIEW

**Comment by not-an-aardvark** [comment]
The tests are failing because travis uses an outdated version of mongodb (2.4) for some reason. I'll figure out how to upgrade that.

**Comment by YaManicKill** [comment]
You'll probably need to do the thing we did with fapp. Travis is terrible
at doing these things automatically. I'll do that.

**Comment by not-an-aardvark** [comment]
I already fixed it, this can be merged whenever.
